File: node_tools/src/validators/tool-validator.ts
import { access } from 'node:fs/promises';
import { constants as FS_CONSTANTS } from 'node:fs';
import { execa } from 'execa';
import which from 'which';
import { Tool, ToolValidationResult, getToolMissingSections } from '../models/tool.js';

export interface ValidationReport {
  totalTools: number;
  validatedTools: number;
  validTools: number;
  invalidTools: number;
  missingTools: number;
  results: ToolValidationResult[];
  summary: {
    existsCount: number;
    executableCount: number;
    hasVersionCount: number;
    completeTools: number;
    averageCompleteness: number;
  };
  errors: string[];
  generatedAt: string;
}

export class ToolValidator {
  private readonly versionCommands = ['--version', '-v', 'version', '-V', '--help'];
  private readonly timeoutMs = 2000;

  async validateTool(tool: Tool): Promise<ToolValidationResult> {
    const errors: string[] = [];
    const warnings: string[] = [];
    
    // Basic validation
    if (!tool.name || tool.name.trim().length === 0) {
      errors.push('Tool name is required');
    }

    if (!tool.description || tool.description.trim().length === 0) {
      errors.push('Tool description is required');
    }

    if (!tool.category || tool.category.trim().length === 0) {
      errors.push('Tool category is required');
    }

    if (tool.difficulty < 1 || tool.difficulty > 5) {
      warnings.push('Tool difficulty should be between 1 and 5 stars');
    }

    // Check completeness
    const completeness = {
      hasDescription: Boolean(tool.description && tool.description.trim().length > 0),
      hasUseCases: Boolean(tool.commonUseCases && tool.commonUseCases.length > 0),
      hasExamples: Boolean(tool.examples && tool.examples.length > 0),
      hasLocation: Boolean(tool.location && tool.location.trim().length > 0),
      score: 0
    };

    const sections = [
      completeness.hasDescription,
      completeness.hasUseCases,
      completeness.hasExamples,
      completeness.hasLocation
    ];
    completeness.score = sections.filter(Boolean).length / sections.length;

    if (completeness.score < 0.75) {
      const missingSections = getToolMissingSections(tool);
      warnings.push(`Tool documentation is incomplete (missing: ${missingSections.join(', ')})`);
    }

    // Check if tool exists in system
    let exists = false;
    let isExecutable = false;
    let hasVersion = false;
    let version: string | undefined;

    try {
      if (tool.location && tool.location.trim().length > 0) {
        exists = await this.checkToolExists(tool.location);
        
        if (exists) {
          isExecutable = await this.checkIsExecutable(tool.location);
          
          if (isExecutable) {
            const versionResult = await this.getToolVersion(tool.location);
            hasVersion = versionResult.hasVersion;
            version = versionResult.version;
          }
        }
      } else {
        warnings.push('Tool location not specified');
      }
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      warnings.push(`Failed to validate tool existence: ${errorMessage}`);
    }

    // Validate examples syntax
    if (tool.examples && tool.examples.length > 0) {
      for (const example of tool.examples) {
        if (!this.validateExampleSyntax(example)) {
          warnings.push('Some examples may have syntax issues');
          break;
        }
      }
    }

    const isValid = errors.length === 0;

    const result: ToolValidationResult = {
      tool,
      isValid,
      exists,
      isExecutable,
      hasVersion,
      errors,
      warnings,
      completeness
    };

    if (version !== undefined) {
      result.version = version;
    }

    return result;
  }

  async validateTools(tools: Tool[]): Promise<ToolValidationResult[]> {
    const results: ToolValidationResult[] = [];
    
    for (const tool of tools) {
      try {
        const result = await this.validateTool(tool);
        results.push(result);
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        results.push({
          tool,
          isValid: false,
          exists: false,
          isExecutable: false,
          hasVersion: false,
          errors: [`Validation failed: ${errorMessage}`],
          warnings: [],
          completeness: {
            hasDescription: false,
            hasUseCases: false,
            hasExamples: false,
            hasLocation: false,
            score: 0
          }
        });
      }
    }

    return results;
  }

  async generateReport(tools: Tool[]): Promise<ValidationReport> {
    const results = await this.validateTools(tools);
    
    const validatedTools = results.length;
    const validTools = results.filter(r => r.isValid).length;
    const invalidTools = results.filter(r => !r.isValid).length;
    const missingTools = results.filter(r => !r.exists).length;
    
    const summary = {
      existsCount: results.filter(r => r.exists).length,
      executableCount: results.filter(r => r.isExecutable).length,
      hasVersionCount: results.filter(r => r.hasVersion).length,
      completeTools: results.filter(r => r.completeness.score >= 0.75).length,
      averageCompleteness: validatedTools > 0 
        ? results.reduce((sum, r) => sum + r.completeness.score, 0) / validatedTools 
        : 0
    };

    const errors = results
      .flatMap(r => r.errors.map(error => `${r.tool.name}: ${error}`))
      .slice(0, 20); // Limit to first 20 errors

    return {
      totalTools: tools.length,
      validatedTools,
      validTools,
      invalidTools,
      missingTools,
      results,
      summary,
      errors,
      generatedAt: new Date().toISOString()
    };
  }

  private async checkToolExists(toolLocation: string): Promise<boolean> {
    try {
      // First try to find it in PATH
      await which(toolLocation);
      return true;
    } catch {
      // If not in PATH, check if it's a file path
      try {
        await access(toolLocation, FS_CONSTANTS.F_OK);
        return true;
      } catch {
        return false;
      }
    }
  }

  private async checkIsExecutable(toolLocation: string): Promise<boolean> {
    try {
      // Check if it's executable using fs.access
      await access(toolLocation, FS_CONSTANTS.X_OK);
      return true;
    } catch {
      // If not a direct path, try to resolve from PATH
      try {
        const resolvedPath = await which(toolLocation);
        await access(resolvedPath, FS_CONSTANTS.X_OK);
        return true;
      } catch {
        return false;
      }
    }
  }

  private async getToolVersion(toolLocation: string): Promise<{ hasVersion: boolean; version?: string }> {
    for (const versionFlag of this.versionCommands) {
      try {
        const result = await this.executeWithTimeout(toolLocation, [versionFlag]);
        
        if (result.stdout || result.stderr) {
          const output = result.stdout || result.stderr;
          const version = this.extractVersionFromOutput(output);
          return {
            hasVersion: true,
            version: version || output.split('\n')[0].trim()
          };
        }
      } catch (error) {
        // Continue to next version command
        continue;
      }
    }

    return { hasVersion: false };
  }

  private async executeWithTimeout(
    command: string, 
    args: string[]
  ): Promise<{ stdout: string; stderr: string }> {
    try {
      const result = await execa(command, args, {
        timeout: this.timeoutMs,
        reject: false
      });
      
      return {
        stdout: result.stdout || '',
        stderr: result.stderr || ''
      };
    } catch (error) {
      if (error instanceof Error && error.message.includes('timed out')) {
        throw new Error('Command timeout');
      }
      throw error;
    }
  }

  private extractVersionFromOutput(output: string): string | undefined {
    const versionRegex = /(?:version|v\.?)\s*:?\s*([0-9]+(?:\.[0-9]+)*(?:-[a-zA-Z0-9]+)*)/i;
    const match = output.match(versionRegex);
    return match?.[1];
  }

  private validateExampleSyntax(example: string | { command: string; description: string }): boolean {
    const command = typeof example === 'string' ? example : example.command;
    
    // Basic syntax validation for common shell patterns
    const invalidPatterns = [
      /\$\s*$/, // Ends with $ prompt
      /^\s*[{}]\s*$/, // Just brackets
      /^[^a-zA-Z0-9_\/\-\.]/, // Starts with invalid characters
    ];

    return !invalidPatterns.some(pattern => pattern.test(command));
  }

  async validateToolsInBatch(
    tools: Tool[], 
    batchSize: number = 10
  ): Promise<ToolValidationResult[]> {
    const results: ToolValidationResult[] = [];
    
    for (let i = 0; i < tools.length; i += batchSize) {
      const batch = tools.slice(i, i + batchSize);
      const batchResults = await Promise.all(
        batch.map(tool => this.validateTool(tool))
      );
      results.push(...batchResults);
    }

    return results;
  }

  getValidationSummary(results: ToolValidationResult[]): string {
    const total = results.length;
    const valid = results.filter(r => r.isValid).length;
    const exists = results.filter(r => r.exists).length;
    const executable = results.filter(r => r.isExecutable).length;
    const hasVersion = results.filter(r => r.hasVersion).length;
    
    const avgCompleteness = total > 0 
      ? Math.round((results.reduce((sum, r) => sum + r.completeness.score, 0) / total) * 100)
      : 0;

    return [
      `Validated ${total} tools:`,
      `• ${valid} valid (${Math.round((valid/total)*100)}%)`,
      `• ${exists} exist in system (${Math.round((exists/total)*100)}%)`,
      `• ${executable} executable (${Math.round((executable/total)*100)}%)`,
      `• ${hasVersion} have version info (${Math.round((hasVersion/total)*100)}%)`,
      `• Average completeness: ${avgCompleteness}%`
    ].join('\n');
  }
}

File: dart_tools/lib/validators/tool_validator.dart
import 'dart:io';
import '../models/tool.dart';

/// Validation result for a tool
class ValidationResult {
  final Tool tool;
  final bool exists;
  final bool isExecutable;
  final String? version;
  final List<String> issues;

  ValidationResult({
    required this.tool,
    required this.exists,
    required this.isExecutable,
    this.version,
    required this.issues,
  });

  bool get isValid => exists && isExecutable && issues.isEmpty;
}

/// Validator for CLI tools
class ToolValidator {
  final bool verbose;

  ToolValidator({this.verbose = false});

  /// Validate a single tool
  Future<ValidationResult> validateTool(Tool tool) async {
    final issues = <String>[];
    
    // Check if tool is complete
    if (!tool.isComplete) {
      issues.addAll(tool.missingSections.map((s) => 'Missing $s section'));
    }

    // Check if tool exists on system
    var exists = false;
    var isExecutable = false;
    String? version;

    if (tool.location.isNotEmpty) {
      final file = File(tool.location);
      exists = await file.exists();
      
      if (exists) {
        final stat = await file.stat();
        isExecutable = (stat.mode & 0x40) != 0; // Check executable bit
        
        // Try to get version
        version = await _getToolVersion(tool.name);
      } else {
        // Try to find in PATH
        final whichResult = await _runCommand('which', [tool.name]);
        if (whichResult != null && whichResult.isNotEmpty) {
          exists = true;
          isExecutable = true;
          version = await _getToolVersion(tool.name);
        }
      }
    }

    if (!exists) {
      issues.add('Tool not found at ${tool.location}');
    }
    
    if (exists && !isExecutable) {
      issues.add('Tool exists but is not executable');
    }

    // Validate examples syntax (basic check)
    for (final example in tool.examples) {
      if (example.trim().isEmpty) continue;
      
      // Check for common syntax issues
      if (example.contains('```') && !example.contains('```bash')) {
        issues.add('Example code block should specify language: ```bash');
      }
      
      // Check for placeholder values that should be replaced
      if (RegExp(r'<[A-Z_]+>|\[.*?\]').hasMatch(example)) {
        // This is OK for examples with placeholders
        continue;
      }
    }

    return ValidationResult(
      tool: tool,
      exists: exists,
      isExecutable: isExecutable,
      version: version,
      issues: issues,
    );
  }

  /// Validate multiple tools
  Future<List<ValidationResult>> validateTools(List<Tool> tools) async {
    final results = <ValidationResult>[];
    
    for (final tool in tools) {
      if (verbose) {
        stdout.write('Validating ${tool.name}...');
      }
      
      final result = await validateTool(tool);
      results.add(result);
      
      if (verbose) {
        if (result.isValid) {
          print(' ✓');
        } else {
          print(' ✗ (${result.issues.length} issues)');
        }
      }
    }
    
    return results;
  }

  /// Get tool version
  Future<String?> _getToolVersion(String toolName) async {
    // Try common version flags
    final versionFlags = ['--version', '-v', 'version', '-V'];
    
    for (final flag in versionFlags) {
      final result = await _runCommand(toolName, [flag]);
      if (result != null && result.isNotEmpty) {
        // Extract version number from output
        final versionMatch = RegExp(r'(\d+\.[\d.]+)').firstMatch(result);
        if (versionMatch != null) {
          return versionMatch.group(1);
        }
        return result.split('\n').first;
      }
    }
    
    return null;
  }

  /// Run a command and return output
  Future<String?> _runCommand(String command, List<String> args) async {
    try {
      final result = await Process.run(
        command,
        args,
        runInShell: true,
      ).timeout(
        const Duration(seconds: 2),
        onTimeout: () => ProcessResult(0, 1, '', ''),
      );
      
      if (result.exitCode == 0) {
        return result.stdout.toString().trim();
      }
    } catch (e) {
      // Command failed or not found
    }
    return null;
  }

  /// Generate validation report
  String generateReport(List<ValidationResult> results) {
    final buffer = StringBuffer();
    
    final totalTools = results.length;
    final validTools = results.where((r) => r.isValid).length;
    final missingTools = results.where((r) => !r.exists).length;
    final incompleteTools = results.where((r) => !r.tool.isComplete).length;
    
    buffer.writeln('='.padRight(60, '='));
    buffer.writeln('Tool Validation Report');
    buffer.writeln('='.padRight(60, '='));
    buffer.writeln();
    buffer.writeln('Summary:');
    buffer.writeln('  Total Tools: $totalTools');
    buffer.writeln('  Valid Tools: $validTools');
    buffer.writeln('  Missing Tools: $missingTools');
    buffer.writeln('  Incomplete Documentation: $incompleteTools');
    buffer.writeln();
    
    if (results.any((r) => !r.isValid)) {
      buffer.writeln('Issues Found:');
      buffer.writeln('-'.padRight(60, '-'));
      
      for (final result in results.where((r) => !r.isValid)) {
        buffer.writeln();
        buffer.writeln('${result.tool.name} (line ${result.tool.lineNumber}):');
        for (final issue in result.issues) {
          buffer.writeln('  - $issue');
        }
      }
    }
    
    return buffer.toString();
  }
}